Implement simple Library Management API

vladyslav-tmf opened the pull request.

vladyslav-tmf committed:
Initial project setup with Poetry

- Initialize Poetry project
- Add FastAPI with all extras
- Add SQLAlchemy as ORM
- Create basic project structure

vladyslav-tmf committed:
Add database configuration

- Set up SQLAlchemy with SQLite database

vladyslav-tmf committed:
Add SQLAlchemy models for Author and Book

- Create Author model with id, name, bio fields and books relationship
- Create Book model with id, title, summary, publication_date fields and author relationship
- Set up one-to-many relationship between Author and Book
- Add string representations for models using __repr__
- Configure appropriate field constraints and types

vladyslav-tmf committed:
Implement Pydantic schemas for API models

- Add base and extended schemas for Book and Author models
- Implement response schemas for paginated lists
- Configure automatic conversion from SQLAlchemy models

vladyslav-tmf committed:
Implement Author management functions

- Add CRUD operations for authors

vladyslav-tmf committed:
Implement Book management functions

- Add private _paginate_query function for unified pagination handling
- Implement create_book function with author existence validation
- Add get_books_with_pagination with optional author filtering
- Refactor pagination logic to use shared helper function

vladyslav-tmf committed:
Fix formatting in crud.py

vladyslav-tmf committed:
Implement API endpoints and error handling

- Create FastAPI application with routers for authors and books
- Add new AuthorListResponse schema for paginated author responses
- Update .gitignore to exclude database files

vladyslav-tmf committed:
Fix formatting

mateacademy-ai-mentor commented on app/crud.py:
```diff
@@ -0,0 +1,132 @@
+    Returns:
+        models.Author: The created author model with database-populated fields.
+    """
+    new_author = models.Author(**author.model_dump())
```
The use of `author.model_dump()` is incorrect as `model_dump()` is not a standard method in Pydantic models. You should use `author.dict()` to convert the Pydantic model to a dictionary.

vladyslav-tmf replied:
The `dict()` method was in Pydantic v1, but since version 2.0 it has been replaced by `model_dump()`.

mateacademy-ai-mentor commented on app/crud.py:
```diff
@@ -0,0 +1,132 @@
+    if not author:
+        return None
+
+    new_book = models.Book(**book.model_dump(), author_id=author_id)
```
Similar to the issue with `author.model_dump()`, `book.model_dump()` should be replaced with `book.dict()` to convert the Pydantic model to a dictionary.

vladyslav-tmf replied:
The `dict()` method was in Pydantic v1, but since version 2.0 it has been replaced by `model_dump()`.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,207 @@
+
+
+@authors_router.post(
+    "/", response_model=schemas.Author, status_code=status.HTTP_201_CREATED
```
Ensure that `schemas.Author` is correctly defined in `schemas.py` to match the expected response model. Verify that it includes all necessary fields and configurations for serialization.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,207 @@
+        )
+
+
+@authors_router.get("/", response_model=schemas.AuthorListResponse)
```
Ensure that `schemas.AuthorListResponse` is correctly defined in `schemas.py` to match the expected response model. Verify that it includes all necessary fields and configurations for serialization.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,207 @@
+    )
+
+
+@authors_router.get("/{author_id}", response_model=schemas.Author)
```
Ensure that `schemas.Author` is correctly defined in `schemas.py` to match the expected response model. Verify that it includes all necessary fields and configurations for serialization.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,207 @@
+
+
+@books_router.post(
+    "/", response_model=schemas.Book, status_code=status.HTTP_201_CREATED
```
Ensure that `schemas.Book` is correctly defined in `schemas.py` to match the expected response model. Verify that it includes all necessary fields and configurations for serialization.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,207 @@
+        )
+
+
+@books_router.get("/", response_model=schemas.BookListResponse)
```
Ensure that `schemas.BookListResponse` is correctly defined in `schemas.py` to match the expected response model. Verify that it includes all necessary fields and configurations for serialization.

mateacademy-ai-mentor commented on app/models.py:
```diff
@@ -0,0 +1,68 @@
+
+    Notes:
+        The model enforces unique constraints for:
+        - Combination of title and author_id.
+        - Combination of title and publication_date.
+        - Combination of title and summary.
```
The unique constraints defined here might be overly restrictive. Typically, a unique constraint on just the combination of `title` and `author_id` should suffice to ensure that an author doesn't have multiple books with the same title. The additional constraints on `title` with `publication_date` and `summary` might not be necessary and could lead to unexpected behavior when adding books with similar titles but different summaries or publication dates.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        The model_config enables automatic conversion from SQLAlchemy models.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
The `model_config` attribute is not a standard Pydantic configuration. If you intend to use Pydantic's ORM mode, you should use `Config` class with `orm_mode = True` instead. This will allow Pydantic to work with ORM objects.

vladyslav-tmf replied:
In Pydantic v2, instead of the `Config` class with `orm_mode = True`, the `model_config` with `from_attributes = True` is used.
Using `model_config = ConfigDict(from_attributes=True)` is the modern and correct approach for Pydantic v2.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        The model_config enables automatic conversion from SQLAlchemy models.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
Similar to the previous comment, replace `model_config` with a `Config` class and set `orm_mode = True` to enable ORM compatibility.

vladyslav-tmf replied:
In Pydantic v2, instead of the `Config` class with `orm_mode = True`, the `model_config` with `from_attributes = True` is used.
Using `model_config = ConfigDict(from_attributes=True)` is the modern and correct approach for Pydantic v2.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        total_items (int): Total number of items across all pages.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
Replace `model_config` with a `Config` class and set `orm_mode = True` if you want to enable ORM compatibility for the `BookListResponse` schema.

vladyslav-tmf replied:
In Pydantic v2, instead of the `Config` class with `orm_mode = True`, the `model_config` with `from_attributes = True` is used.
Using `model_config = ConfigDict(from_attributes=True)` is the modern and correct approach for Pydantic v2.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        total_items (int): Total number of items across all pages.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
Replace `model_config` with a `Config` class and set `orm_mode = True` if you want to enable ORM compatibility for the `AuthorListResponse` schema.

vladyslav-tmf replied:
In Pydantic v2, instead of the `Config` class with `orm_mode = True`, the `model_config` with `from_attributes = True` is used.
Using `model_config = ConfigDict(from_attributes=True)` is the modern and correct approach for Pydantic v2.

mateacademy-ai-mentor reviewed (changes requested):
Thank you for your hard work on this project! Your implementation is quite comprehensive and meets many of the task requirements. However, there are a few critical issues that need to be addressed before we can approve the pull request:

1. **Pydantic Configuration**: The use of `model_config` is incorrect. Instead, you should define a `Config` class within your Pydantic models and set `orm_mode = True` to enable ORM compatibility. This change is necessary for the correct serialization of ORM objects.

2. **Unique Constraints**: The unique constraints in your models might be too restrictive. Typically, a unique constraint on the combination of `title` and `author_id` should be sufficient. The additional constraints could lead to unexpected behavior and should be reconsidered.

3. **Model Serialization**: The use of `model_dump()` is not standard in Pydantic. You should replace it with `dict()` for converting Pydantic models to dictionaries.

These changes are crucial for ensuring the functionality and correctness of your API. Once these issues are resolved, your code will be in great shape! Keep up the good work, and don't hesitate to reach out if you need further assistance. 😊

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

🔄 Fix the issues, commit the changes, and then re-request my review using the `Re-request review` button in the Reviewers tab.

vladyslav-tmf committed:
Refactor Book model constraints

- Remove unique constraints on publication_date and summary fields in the Book model
- Update documentation to reflect the changes in unique constraints

mateacademy-ai-mentor commented on app/crud.py:
```diff
@@ -0,0 +1,132 @@
+    Returns:
+        models.Author: The created author model with database-populated fields.
+    """
+    new_author = models.Author(**author.model_dump())
```
The use of `model_dump()` is incorrect for Pydantic models. You should use `dict()` to convert the Pydantic model to a dictionary. This is necessary for proper serialization.

mateacademy-ai-mentor commented on app/crud.py:
```diff
@@ -0,0 +1,132 @@
+    if not author:
+        return None
+
+    new_book = models.Book(**book.model_dump(), author_id=author_id)
```
The use of `model_dump()` is incorrect for Pydantic models. You should use `dict()` to convert the Pydantic model to a dictionary. This is necessary for proper serialization.

mateacademy-ai-mentor commented on app/models.py:
```diff
@@ -0,0 +1,64 @@
+    __tablename__ = "authors"
+
+    id: Mapped[int] = mapped_column(primary_key=True)
+    name: Mapped[str] = mapped_column(String(100), unique=True)
```
The unique constraint on the `name` field of the `Author` model might be too restrictive. Consider whether this is necessary, as it could lead to unexpected behavior if two authors share the same name.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        The model_config enables automatic conversion from SQLAlchemy models.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
The use of `model_config` is incorrect. Define a `Config` class within the `Book` model and set `orm_mode = True` to enable ORM compatibility.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        The model_config enables automatic conversion from SQLAlchemy models.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
The use of `model_config` is incorrect. Define a `Config` class within the `Author` model and set `orm_mode = True` to enable ORM compatibility.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        total_items (int): Total number of items across all pages.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
The use of `model_config` is incorrect. Define a `Config` class within the `BookListResponse` model and set `orm_mode = True` to enable ORM compatibility.

mateacademy-ai-mentor commented on app/schemas.py:
```diff
@@ -0,0 +1,120 @@
+        total_items (int): Total number of items across all pages.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
The use of `model_config` is incorrect. Define a `Config` class within the `AuthorListResponse` model and set `orm_mode = True` to enable ORM compatibility.

mateacademy-ai-mentor reviewed (approved):
****

Great effort on your submission! 🎉 While there are a few issues to address, your work is on the right track, and I'm approving this pull request to help you move forward with your learning. Here are some key points to focus on for improvement:

1. **Pydantic Configuration**: The use of `model_config` is incorrect. You should define a `Config` class within each Pydantic model (`AuthorListResponse`, `BookListResponse`, `Author`, and `Book`) and set `orm_mode = True` to enable ORM compatibility. This is crucial for the correct serialization of ORM objects.

2. **Unique Constraints**: The unique constraint on the `name` field of the `Author` model might be too restrictive. Consider whether this is necessary, as it could lead to unexpected behavior if two authors share the same name. Typically, a unique constraint on the combination of `title` and `author_id` in the `Book` model is sufficient.

3. **Model Serialization**: The use of `model_dump()` is incorrect for Pydantic models. You should use `dict()` to convert the Pydantic model to a dictionary. This change is necessary for proper serialization in your CRUD operations.

Please review these comments and make the necessary adjustments to improve your code. Addressing these issues will help you avoid similar problems in the future and enhance your understanding of Pydantic and SQLAlchemy. Keep up the good work, and continue learning from your experiences! 🚀

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

vladyslav-tmf committed:
Enhance pagination in authors and books endpoints

- Add a private helper function to construct pagination query parameters for authors and books
- Replace inline URL construction with "urlencode" for better handling of query parameters

vladyslav-tmf committed:
Fix formatting

AnyoneClown reviewed (approved):
LGTM
